### PR TITLE
Add PostgreSQL database backend for naffka

### PR DESCRIPTION
**Pull request by NegativeMjark**

### CODE REVIEW

File: postgresqldatabase.go [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+import (
+	"database/sql"
+	"sync"
+	"time"
+)
+
+const postgresqlSchema = `
+CREATE SEQUENCE IF NOT EXISTS naffka_topic_nid_seq;
+CREATE TABLE IF NOT EXISTS naffka_topics (
+	topic_name TEXT PRIMARY KEY,
```
Comment (richvdh) [inline postgresqldatabase.go]: shouldn't the `nid` be the primary key?

File: postgresqldatabase.go [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+	topicNID, err = p.getTopicNID(txn, topicName)
+	if err != nil {
+		return 0, err
+	}
+	if topicNID != 0 {
+		return topicNID, err
+	}
+	err = txn.Stmt(p.insertTopicStmt).QueryRow(topicName).Scan(&topicNID)
+	if err == sql.ErrNoRows {
+		return p.getTopicNID(txn, topicName)
```
Comment (richvdh) [inline postgresqldatabase.go]: I'm a bit confused about why both this and `ON CONFLICT DO NOTHING RETURNING (topic_nid)` are necessary. Cn you explain?

File: postgresqldatabase.go [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+	})
+}
+
+// FetchMessages implements Database.
+func (p *postgresqlDatabase) FetchMessages(topic string, startOffset, endOffset int64) (messages []Message, err error) {
+	topicNID, err := p.getTopicNID(nil, topic)
+	if err != nil {
+		return
+	}
+	rows, err := p.selectMaxOffsetStmt.Query(topicNID, startOffset, endOffset)
```
Comment (richvdh) [inline postgresqldatabase.go]: selectMessagesStmt ?

**Review by richvdh — CHANGES_REQUESTED** [review]

File: postgresqldatabase.go (hunk 1 above)
Comment (NegativeMjark) [reply postgresqldatabase.go]: This is a table to lookup topic_name to topic_nid, we don't actually need to go the other way.

File: postgresqldatabase.go (hunk 3 above)
Comment (NegativeMjark) [reply postgresqldatabase.go]: oops, fixed in 671be41

File: postgresqldatabase.go (hunk 2 above)
Comment (NegativeMjark) [reply postgresqldatabase.go]: Yep, I've explained in 671be41 and 6612b90

File: postgresqldatabase.go [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+	message_key BYTEA NOT NULL,
+	message_value BYTEA NOT NULL,
+	message_timestamp_ns BIGINT NOT NULL,
+	UNIQUE (topic_nid, message_offset)
+);
+`
+
+const insertTopicSQL = "" +
+	"INSERT INTO naffka_topics (topic_name) VALUES ($1)" +
+	" ON CONFLICT DO NOTHING RETURNING (topic_nid)"
```
Comment (richvdh) [inline postgresqldatabase.go]: ohhhh, this isn't `ON CONFLICT (DO NOTHING RETURNING (topic_nid))`, but rather `(ON CONFLICT DO NOTHING) (RETURNING (topic_nid))`.

I think breaking this into two lines would have been less likely to confuse me.

**Review by richvdh — APPROVED** [review]
lgtm, especially if you add a linebreak in `insertTopicSQL`

**Pull request merged**